### ISSUE TITLE
Create terminology MRML node

### ISSUE THREAD

**Issue by cpinter** [Slicer/Slicer#6726]
## Is your feature request related to a problem? Please describe.

When saving a scene that contains a segmentation using a custom terminology, the scene does not contain the custom terminology. As a consequence, the scene cannot be loaded just by loading the MRB in a freshly started Slicer, but first the user needs to load the JSON and then load the MRB.

## Describe the solution you'd like

Having a terminology MRML node that saves/loads itself into/from the mrml file and saves the json along the scene would solve the problem.

## Describe alternatives you've considered

I don't think there is another solution that makes sense. Slicer stores data in MRML nodes, and a terminology context is data. So if we want to save it, then it should happen in a MRML node.

## Additional context

The terminologies are currently stored in the internal class of the terminology logic, see
https://github.com/Slicer/Slicer/blob/5a3f3381d16c7151b695d2597695951869c9981d/Modules/Loadable/Terminologies/Logic/vtkSlicerTerminologiesModuleLogic.cxx#L159

**Comment by lassoan** [comment]
> the scene does not contain the custom terminology

The segmentation node contains the full terminology (context name, category, type, modifier, etc. codes). Is the problem that when you load this segmentation and you open the terminology selector then you don't see the term in the list?

**Comment by cpinter** [comment]
Well, yes. But on a more basic level, the saved then loaded scene does not contain an important data object that Slicer contained when saving.

**Comment by lassoan** [comment]
It is similar to how colormaps are managed. Only those colormaps (terms) are saved into the scene that are used.

Does it cause you any problem that some terms are not available in the terminology list?

**Comment by cpinter** [comment]
> Only those colormaps (terms) are saved into the scene that are used.

I think that this is not true, and no terminology is saved with the scene no matter what is used. Couldn't even be saved, because there is no node for it that could contain it.

> Does it cause you any problem that some terms are not available in the terminology list?

You load a scene that had for example the spine terminology (all vertebrae, all discs, etc), and the terminology information cannot be seen any more. When you double-click the color, it says None.

I have a hard time understanding why it would not be a problem that you save a scene, and when you load it, an important component does not work any more. Can you please give some background about your reasoning?

**Comment by lassoan** [comment]
Custom color maps are saved with the scene. Terminology is saved into the segmentation file.

> You load a scene that had for example the spine terminology (all vertebrae, all discs, etc), and the terminology information cannot be seen any more.

Yes, this is a problem. This could be either fixed in the GUI (allow adding/editing codes), or we could add missing codes to the terminology before opening the selector widget.

> I have a hard time understanding why it would not be a problem that you save a scene, and when you load it, an important component does not work any more.

You can load, edit, and save the segmentation and all information is preserved. You can see the full terminology information in the tooltip of the segment list widget. There is an issue with the terminology selector widget.

We could store the terminology in color table node or in a new terminology node, which would be definitely useful (you could delete a segment with a custom code and the code would not be lost, etc.). But it would be probably more work than fixing the terminology selector widget, and we need to fix the widget anyway (because someone can load a segmentation file that contains cus om codes without an accompanying terminology node).

**Comment by cpinter** [comment]
> But it would be probably more work than fixing the terminology selector widget, and we need to fix the widget anyway (because someone can load a segmentation file that contains custom codes without an accompanying terminology node).

I agree with this. We could focus on fixing the widget first, and making sure any terminology entry that is being edited is added if is not contained in the current context (this may be a question, probably better if we add a new, temporary context for the missing entries instead of adding it to an existent one).

A node would be useful, and it does not seem to be a big work on first glance, but probably after digging a bit it could lead far.
